dockstore-webservice/src/main/java/io/dockstore/webservice/permissions/sam/SamPermissionsImpl.java

```diff
@@ -39,8 +39,54 @@
 import org.slf4j.LoggerFactory;
 
 /**
- * An implementation of the {@link PermissionsInterface} that makes
- * calls to SAM.
+ * <p>An implementation of the {@link PermissionsInterface} that makes
+ * calls to SAM.</p>
+ *
+ * <p>
+ *     Calls to SAM use a Google access token. This token can be acquired in 3 ways:
+ *     <ul>
+ *         <li>Using the user's Google access token in the database.</li>
+ *         <li>Minting and using a new Google access token with the user's refresh token in the
+ *         database. This is only done when we've detected the Google access token has expired.</li>
+ *         <li>A call from the "outside", e.g., Terra accesses a Dockstore API with a Google access
+ *         token minted in Terra. In this case, the access token is stored in the transient
+ *         <code>User.temporaryCredential</code> field. Note that this use case is technically
+ *         possible but not currently used by Terra. Also note that we check the token's "audience"
+ *         to ensure only approved Google clients can access Dockstore this way.</li>
+ *     </ul>
+ * </p>
+ *
+ * <p>What can go wrong with auth?
+ * <ul>
+ *     <li>The user's Google account is not registered in Terra/SAM</li>
+ *     <li>The user's Google account in Terra/SAM has been disabled</li>
+ *     <li>The user's Google refresh token has expired. We are unable to mint a new access token
+ *     for the user. The user must relink their Google account in Dockstore to get a new refresh
```

Comment:
Another thing that can go wrong: the user has never linked a Google account (on purpose), which, historically, would sometimes cause the "relink" exception to be thrown:
https://github.com/dockstore/dockstore/blob/be849e3aad4fcf06246e3b8743bfc2eec455bd82/dockstore-webservice/src/main/java/io/dockstore/webservice/permissions/sam/SamPermissionsImpl.java#L542
We have made some previous changes that [mostly?] prevent that from happening, but this PR would be a great opportunity to make sure...

Reply:
Good call, added some tests and tweaked code